usersession/xdgopenproxy/portal_launcher.go

```diff
@@ -133,36 +169,91 @@ func (p *portalLauncher) portalCall(bus *dbus.Conn, call func() (dbus.ObjectPath
 	}
 }
 
-func (p *portalLauncher) OpenFile(bus *dbus.Conn, filename string) error {
-	portal, err := p.desktopPortal(bus)
+func (p *portalLauncher) OpenFile(bus *dbus.Conn, filename string, options LauncherModifiers) error {
+
+	var attemptWritable bool
+	var fd int
+
+	pathStat, err := os.Stat(filename)
+	if err != nil {
+		return &responseError{msg: err.Error()}
+	}
+
+	// Trying to make paths writable is not valid, but doesn't fail here, so explicitly avoid bothering.
+	// Otherwise try open a file with RDRW and if that fails, try again with RDONLY.
+	if pathStat.IsDir() {
+		fd, err = syscall.Open(filename, syscall.O_RDONLY, 0)
+		if err != nil {
+			return &responseError{msg: err.Error()}
+		}
+	} else {
+		attemptWritable = true
+		fd, err = syscall.Open(filename, syscall.O_RDWR, 0)
+		if err != nil {
+			fd, err = syscall.Open(filename, syscall.O_RDONLY, 0)
+			if err != nil {
+				return &responseError{msg: err.Error()}
+			}
+			attemptWritable = false
+		}
+	}
+	defer syscall.Close(fd)
+
+	portal, err := p.documentsPortal(bus)
 	if err != nil {
 		return err
 	}
 
-	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
+	// Try add the file to the documents portal, allowing OpenURI to mediate sandbox permissions if needed.
+	// Failure here is not a critical error, the application opened by OpenURI might not need mediation.
+	if attemptWritable {
+		var docID interface{}
+		err = portal.Call(documentsPortalDocumentsIface+".Add", 0, dbus.UnixFD(fd), false, false).Store(&docID)
+		if err != nil {
+			attemptWritable = false
+		}
+	}
```

Comment:
Is this bit actually necessary? I would have thought the combination of an O_RDWR file descriptor and writable=true option would be enough.

Reply:
Yes, according to the specifications I've read and testing, the file must be registered by the documents portal or writable doesn't have the proper effect.

https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.OpenURI
"This key only takes effect the uri points to a local file that is exported in the document portal"

Reply:
You've trimmed off the end of that sentence:

> This key only takes effect the uri points to a local file that is exported in the document portal, and the chosen application is sandboxed itself.

My reading is that it is saying that the option will only affect whether the file is writable if (a) the application requires sandboxing, and (b) it was necessary to use the document portal to make the file available to the app. So as an example, if you had one snap asking to open a file in the user's home directory handled by another snap and both snaps plugged the home interface, the second snap would be able to write to the file no matter what writable was set to.

Reply:
I agree with that example, if both apps have access without mediation then the writable flag isn't important and you could consider the documents portal call wasted effort. However, if the recieving app doesn't have access to the file and writable is set, it won't actually work unless the file was added to the documents portal in advance, which is why the Add call is done for every file because it's impossible to know in advance whether it might be needed.

Reply:
That's not what the code does though. If you pass writable=true, two things happen:

1. xdg-desktop-portal will check to make sure that the file descriptor passed in was opened with write permissions (to prove that the caller can write to the file)
2. if the target app is sandboxed, it will register the file with the document portal granting the target app write permissions. Note that the document portal might return the original file path if it believes the target app can access the path already. In that case, the writable flag has no effect.

You can't perform (2) from the caller side, since you don't know which app will handle the file.

Reply:
So I must apologise, having recompiled the patch without the highlighted part, you're correct and the above code is unnecessary.

When I said I'd tested it earlier in the comment chain, it was done before I was aware of the snap specific fixes in XDG 1.8.1, as a result of this and some other changes at the time I'd incorrectly assumed that it was a requirement to export to the document portal in the callers context first because it got mixed in with some other incidental changes, combined with my interpretation of the documentation it happened to all fit together.